### PR TITLE
howto: add sandbox config annotation howto

### PR DESCRIPTION
**Pull request by bergwolf**
It explains the details of current supported annotations.

Fixes: #486
Fixes: #294
Signed-off-by: Peng Tao <bergwolf@hyper.sh>

### CODE REVIEW

**Comment by bergwolf** [comment]
Spell check doesn't understand some tech words, what shall I do to fix it?
```
WARNING: Word ''drop': did you mean one of the following?: drop, droop, drupe, troop
WARNING: Word 'MiB': did you mean one of the following?: MB, Mia, Min, Mir, Mil, Mic, Mi, Mis, Mid, Mi B, Nib, Mob, Rib
WARNING: Word 'MiB': did you mean one of the following?: MB, Mia, Min, Mir, Mil, Mic, Mi, Mis, Mid, Mi B, Nib, Mob, Rib
WARNING: Word 'accpets': did you mean one of the following?: accepts, accents, carpets, accosts, aspects, accepter, accepted
WARNING: Word 'dev': did you mean one of the following?: Dev, deb, div, den, rev, deg, def, dew, Nev, Rev, Lev, Kev, devil, devise
WARNING: Word 'dev': did you mean one of the following?: Dev, deb, div, den, rev, deg, def, dew, Nev, Rev, Lev, Kev, devil, devise
WARNING: Word 'enotes': did you mean one of the following?: emotes, notes, denotes, e notes, endnote, notates, denote, note's, entice, endorse
WARNING: Word 'in'': did you mean one of the following?: in, ins, inn, int, inc, ind, inf, ink, In, IN, Ian, ion's
WARNING: Word 'kata': did you mean one of the following?: Kata, data, Nata, Mata, Katina, Katmai
WARNING: Word 'kata': did you mean one of the following?: Kata, data, Nata, Mata, Katina, Katmai
WARNING: Word 'kata': did you mean one of the following?: Kata, data, Nata, Mata, Katina, Katmai
WARNING: Word 'kata': did you mean one of the following?: data, Katz, Kate, Tanaka, attack, kart, Kurt
WARNING: Word 'modprobe': did you mean one of the following?: mod probe, mod-probe, prober, problem, mobbed, mobber, probated, probates
WARNING: Word 'msize': did you mean one of the following?: maize, size, m size, seize, Mozes, mises
WARNING: Word 'nvdimm': did you mean one of the following?: dimming, Nevadian, Nevada
WARNING: Word 'nvdimm': did you mean one of the following?: dimming, Nevadian, Nevada
WARNING: Word 'oci': did you mean one of the following?: loci, foci, sci, obi, cocci, socio, Ci, ocean, Oise
WARNING: Word 'scsi': did you mean one of the following?: SCSI
WARNING: Word 'seccomp': did you mean one of the following?: sec comp, sec-comp, decompose, compose, recompense, comprise, scamper, scampi
WARNING: Word 'seccomp': did you mean one of the following?: sec comp, sec-comp, decompose, compose, recompense, comprise, scamper, scampi
WARNING: Word 'seccomp': did you mean one of the following?: sec comp, sec-comp, decompose, compose, recompense, comprise, scamper, scampi
WARNING: Word 'urandom': did you mean one of the following?: random, u random, Durand, Armando, Omdurman, Brando, grandam, grandma
WARNING: Word 'vcpus': did you mean one of the following?: CPUs, pus, copious, scops
WARNING: Word 'vhost': did you mean one of the following?: VHOST
WARNING: Word 'vhost': did you mean one of the following?: host, ghost, v host, Host, Horst, hoist
WARNING: Word 'virtiofs': did you mean one of the following?: virtuoso, vision's, vivisection
WARNING: Word 'vm': did you mean one of the following?: cm, v, m, vim, em, vs, vi, am, nm, rm, om, um, gm, mm, pm
WARNING: Word 'vsock': did you mean one of the following?: sock, v sock, viscus, viscose
ERROR: Spell check failed for file: 'how-to/how-to-set-sandbox-config-kata.md'
ERROR: spell check failed for document how-to/how-to-set-sandbox-config-kata.md
```

File: how-to/how-to-set-sandbox-config-kata.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+# Kata Configuration Annotations
+There are several kinds of kata configurations and they are listed below.
+
+## Global Options
+| Key | Value Type | Comments |
+|-------| ----- | ----- |
+| `io.katacontainers.config_path` | string | kata config file location that overrides the default config paths |
+| `io.katacontainers.pkg.oci.bundle_path` | string | oci bundle path |
+| `io.katacontainers.pkg.oci.container_type`| string | oci container type. Only accpets `pod_container` and `pod_sandbox` |
```
Comment (grahamwhaley) [inline how-to/how-to-set-sandbox-config-kata.md]: s/accpets/accepts/

File: how-to/how-to-set-sandbox-config-kata.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+## Runtime Options
+| Key | Value Type | Comments |
+|-------| ----- | ----- |
+| `io.katacontainers.config.runtime.disable_guest_seccomp`| boolean | determines if seccomp should be applied inside guest |
```
Comment (grahamwhaley) [inline how-to/how-to-set-sandbox-config-kata.md]: s/seccomp/`seccomp`/ ?

File: how-to/how-to-set-sandbox-config-kata.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+| `io.katacontainers.config.runtime.sandbox_cgroup_only`| boolean | determines if kata processes are managed only in sandbox cgroup |
```
Comment (grahamwhaley) [inline how-to/how-to-set-sandbox-config-kata.md]: s/kata/Kata/ throughout?

File: how-to/how-to-set-sandbox-config-kata.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+| `io.katacontainers.config.runtime.experimental` | boolean | determines if experimental features enabled |
+| `io.katacontainers.config.runtime.internetworking_model` | string| determines how the VM should be connected to the container network interface. Valid values are `macvtap`, `tcfilter` and `none` |
+| `io.katacontainers.config.runtime.disable_new_netns` | boolean | determines if create a netns for hypervisor process |
```
Comment (grahamwhaley) [inline how-to/how-to-set-sandbox-config-kata.md]: reword a little maybe:
> determines if a new netns is created for the hypervisor process

File: how-to/how-to-set-sandbox-config-kata.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+|-------| ----- | ----- |
+| `io.katacontainers.config.agent.kernel_modules` | string | the list of kernel modules and their parameters that will be loaded in the guest kernel. Semicolon separated list of kernel modules and their parameters. These modules will be loaded in the guest kernel using modprobe(8). E.g., `e1000e InterruptThrottleRate=3000,3000,3000 EEE=1; i915 enable_ppgtt=0` |
+| `io.katacontainers.config.agent.enable_tracing` | boolean | enable tracing for the agent |
+| `io.katacontainers.config.agent.trace_mode` | string | the trace mode for the agent |
+| `io.katacontainers.config.agent.trace_type` | string | the trace type for the agent |
+
+## Hypervisor Options
+| Key | Value Type | Comments |
+|-------| ----- | ----- |
+| `io.katacontainers.config.hypervisor.kernel` | string | the kernel needed to boot the container VM |
```
Comment (grahamwhaley) [inline how-to/how-to-set-sandbox-config-kata.md]: s/needed/used/

File: how-to/how-to-set-sandbox-config-kata.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+| `io.katacontainers.config.hypervisor.asset_hash_type` | string | the hash type used for assets verification, default is `sha512` |
+| `io.katacontainers.config.hypervisor.kernel_params` | string | additional guest kernel parameters |
+| `io.katacontainers.config.hypervisor.machine_type` | string | the type of machine being emulated by the hypervisor |
+| `io.katacontainers.config.hypervisor.machine_accelerators` | string | machine specific accelerators for the hypervisor |
+| `io.katacontainers.config.hypervisor.disable_vhost_net` | boolean | specify if vhost-net is not available on the host |
+| `io.katacontainers.config.hypervisor.guest_hook_path` | string | the path within the VM that will be used for 'drop-in' hooks |
+| `io.katacontainers.config.hypervisor.use_vsock` | boolean | specify use of vsock for agent communication |
+| `io.katacontainers.config.hypervisor.hotplug_vfio_on_root_bus` | boolean | indicate if devices need to be hotplugged on the root bus instead of a bridge|
+| `io.katacontainers.config.hypervisor.entropy_source` | string| the path to a host source of entropy (/dev/random, /dev/urandom or real hardware RNG device) |
```
Comment (grahamwhaley) [inline how-to/how-to-set-sandbox-config-kata.md]: s?/dev/random?`/dev/random`?

**Comment by jodh-intel** [comment]
Hi @bergwolf - tal at https://github.com/kata-containers/tests/tree/master/cmd/check-spelling/data/ - all acronymns / abbreviations need to use the "official" spelling (including capitalisation) and yes, s/kata/Kata/ please ;)

You can "bypass" the spell-checker by putting a word in backticks, but that should only be used for things like filenames and command names (see https://github.com/kata-containers/documentation/blob/master/Documentation-Requirements.md#files-and-command-names).

File: how-to/how-to-set-sandbox-config-kata.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
+# Per-Pod Kata Configurations
+
+Kata Containers gives users freedom to customise at per-pod level, by setting
+a wide range of kata specific annotations in the pod spec.
```
Comment (jodh-intel) [inline how-to/how-to-set-sandbox-config-kata.md]: - s/spec/specification/
- s/kata/Kata/

File: how-to/how-to-set-sandbox-config-kata.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+# Kata Configuration Annotations
+There are several kinds of kata configurations and they are listed below.
+
+## Global Options
+| Key | Value Type | Comments |
+|-------| ----- | ----- |
+| `io.katacontainers.config_path` | string | kata config file location that overrides the default config paths |
+| `io.katacontainers.pkg.oci.bundle_path` | string | oci bundle path |
```
Comment (jodh-intel) [inline how-to/how-to-set-sandbox-config-kata.md]: s/oci/OCI/

File: how-to/how-to-set-sandbox-config-kata.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+| `io.katacontainers.config.agent.kernel_modules` | string | the list of kernel modules and their parameters that will be loaded in the guest kernel. Semicolon separated list of kernel modules and their parameters. These modules will be loaded in the guest kernel using modprobe(8). E.g., `e1000e InterruptThrottleRate=3000,3000,3000 EEE=1; i915 enable_ppgtt=0` |
+| `io.katacontainers.config.agent.enable_tracing` | boolean | enable tracing for the agent |
+| `io.katacontainers.config.agent.trace_mode` | string | the trace mode for the agent |
+| `io.katacontainers.config.agent.trace_type` | string | the trace type for the agent |
+
+## Hypervisor Options
+| Key | Value Type | Comments |
+|-------| ----- | ----- |
+| `io.katacontainers.config.hypervisor.kernel` | string | the kernel needed to boot the container VM |
+| `io.katacontainers.config.hypervisor.image` | string | the guest image that will run in the container VM |
```
Comment (jodh-intel) [inline how-to/how-to-set-sandbox-config-kata.md]: Can you sort the rows alphabetically by column 1 (the key)?

**Comment by jodh-intel** [comment]
Just a reminder that this PR needs updating based on review feedback.

**Comment by egernst** [comment]
This is a great document @bergwolf.  I know these edits can be painful, but will you get a chance to update the document?  I'd love to be able to reference this!

**Comment by jodh-intel** [comment]
Ping @bergwolf - just a few `sed`-type updates and this is good to go :smile:

**Comment by bergwolf** [comment]
@grahamwhaley @jodh-intel Sorry for the delay. I've updated the PR to address all comments including sorting all options alphabetically. PTAL.

**Comment by bergwolf** [comment]
The spell checker is still failing but doesn't show any errors. @jodh-intel Any clue? I can't reproduce it locally. `.ci/static-checks.sh` passed successfully on my local machine.

```
INFO: Checking document code blocks
INFO: processing file 'how-to/how-to-set-sandbox-config-kata.md'
INFO: processing file 'how-to/README.md'
INFO: processing file 'how-to/how-to-set-sandbox-config-kata.md'
INFO: processing file 'how-to/README.md'
INFO: processing file 'how-to/how-to-set-sandbox-config-kata.md'
INFO: processing file 'how-to/README.md'
INFO: Checking all document URLs
INFO: Checking URL https://github.com/kata-containers/documentation/wiki/Initial-release-of-Kata-Containers-with-Firecracker-support
INFO: Spell checking file 'how-to/how-to-set-sandbox-config-kata.md'
INFO: Spell check successful for file: 'how-to/how-to-set-sandbox-config-kata.md'
INFO: Spell checking file 'how-to/README.md'
INFO: Spell check successful for file: 'how-to/README.md'
INFO: Spell checking file 'how-to/how-to-set-sandbox-config-kata.md'
INFO: Spell check successful for file: 'how-to/how-to-set-sandbox-config-kata.md'
INFO: Spell checking file 'how-to/README.md'
INFO: Spell check successful for file: 'how-to/README.md'
INFO: Spell checking file 'how-to/how-to-set-sandbox-config-kata.md'
INFO: Spell check successful for file: 'how-to/how-to-set-sandbox-config-kata.md'
INFO: Spell checking file 'how-to/README.md'
INFO: Spell check successful for file: 'how-to/README.md'
The command ".ci/static-checks.sh" failed and exited with 1 during .
```

**Comment by grahamwhaley** [comment]
@bergwolf - yep, working on it https://github.com/kata-containers/tests/pull/2240
we (I) broke it with a PR at the end of last week....

**Comment by bergwolf** [comment]
Thanks @grahamwhaley.

**Comment by GabyCT** [comment]
@bergwolf could you please resolve the conflicts ? thanks

**Pull request merged**